x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termsenum/action/TransportTermsEnumAction.java

```diff
@@ -431,21 +432,29 @@ private boolean canAccess(
 
                 // Current user has potentially many roles and therefore potentially many queries
                 // defining sets of docs accessible
-                Set<BytesReference> queries = indexAccessControl.getDocumentPermissions().getQueries();
-                for (BytesReference querySource : queries) {
-                    QueryBuilder queryBuilder = DLSRoleQueryValidator.evaluateAndVerifyRoleQuery(
-                        querySource,
-                        scriptService,
-                        queryShardContext.getParserConfig().registry(),
-                        securityContext.getUser()
-                    );
-                    QueryBuilder rewrittenQueryBuilder = Rewriteable.rewrite(queryBuilder, queryShardContext);
-                    if (rewrittenQueryBuilder instanceof MatchAllQueryBuilder) {
-                        // One of the roles assigned has "all" permissions - allow unfettered access to termsDict
-                        return true;
+                final List<Set<BytesReference>> listOfRawQueries = indexAccessControl.getDocumentPermissions().getListOfQueries();
+
+                return listOfRawQueries.stream().allMatch(queries -> {
+                    for (BytesReference querySource : queries) {
+                        QueryBuilder queryBuilder = DLSRoleQueryValidator.evaluateAndVerifyRoleQuery(
+                            querySource,
+                            scriptService,
+                            queryShardContext.getParserConfig().registry(),
+                            securityContext.getUser()
+                        );
+                        QueryBuilder rewrittenQueryBuilder;
+                        try {
+                            rewrittenQueryBuilder = Rewriteable.rewrite(queryBuilder, queryShardContext);
+                        } catch (IOException e) {
+                            throw new UncheckedIOException(e);
+                        }
+                        if (rewrittenQueryBuilder instanceof MatchAllQueryBuilder) {
+                            // One of the roles assigned has "all" permissions - allow unfettered access to termsDict
+                            return true;
```

Comment:
does the `allMatch` for the stream change the method's return value from needing ONE of the queries to be MatchAllQueryBuilder to ALL of the queries ?  Is that functionally correct, and should the comment be updated ?

Reply:
This is a [bug fix](https://github.com/elastic/elasticsearch/pull/91151#discussion_r1006622490). So the behaviour is not identical to existing code in all cases:

* If the authenticating subject is *not* an API key, e.g. a normal user, the behaviour is unchanged.
* But if it is an API key, the behaviour is different. Because the existing code only checks whether unfiltered access is allowed by the key's role. This is wrong. The change here checks both key's role and owner user's role. 

I added a new block of comments to the begining of the allMatch loop.

Reply:
Discussed in person, and all is good. I missed that we now have an outer iteration (allMatch) as well as the inner iteration (for) and confused myself. The outer requires all to be true, but that is a function of the the inner only requires only 1 match all. Thanks for helping to clear this up.